### PR TITLE
Introduce React Context for campaign state management

### PR DESCRIPTION
**Pull request by miazima**
Create a context to maintain the ‘global’ state of the current campaign during the create-campaign flow. 

### Improvements
- Removes the need for passing `onNext` down to channel-specific components
   - Can update the campaign state from any component subscribed to context
   - Components will update state of campaign state in component e.g. `setCampaignContext({ ...campaign, subject, body })`
- Removes the need for` finishLaterCallbackRef`
   - finishLaterCallbackRef was used to save current state of unsaved draft messages when finish later button was clicked
   - Instead use a finish later callback state, e.g. `const [finishLaterCallback, setFinishLaterCallback] = useState(null)`
- Removes the need to pass campaign props from Create component to <Channel>Create component, to all channel specific components - passing bulky props down two levels of component tree
   - Campaign state is shared across all channel components 

Part of efforts for #795

### Features
- Campaign context for campaign state management
- FinishLater context for finish later callback state management

### Considerations

Components subscribed to a context will re-render for every update to context values, hence resulting in unnecessary re-renders after context value changes.

- Created a separate context for finish later callback instead of including in campaign context
  - only re-renders specified subscribers
- From manually observing the performance of the app, there seemed to be no difference in performance. 
- To assess the changes on the render level, I used react dev tools profiler to profile the campaign creation flow events. Here is a comparison of before and after of the number of renders during an SMS campaign creation flow. From the profiler results, there is no significant increase in renders after introducing context.
https://docs.google.com/document/d/1TOVil0bsWdsftlPoGqTnsRwdpTRS7W_V1mU712exAOA/edit?usp=sharing

### CODE REVIEW

**Review by lamkeewei — CHANGES_REQUESTED** [review]
Accessing of campaign state looks much cleaner now! 👍  

Was thinking if we can combine the functionality of `FinishLaterModalContext` into `CampaignContext`. 

Something along the of lines of just calling `campaignContext.saveDraft()` instead of having to set the `finishLaterContent` in each component and then clearing it on unmount might be cleaner. But I think that will require us to move `handleSaveTemplate` to the `CampaignContext` since the the logic is currently defined in each of the Template component. Perhaps something to think about when we further refactor. 

Other than that just need to resolve the merge conflicts. I think there's a new `onPrevious` method added.

File: frontend/src/components/dashboard/create/telegram/TelegramCredentials.tsx [hunk 1]
```diff
@@ -21,14 +28,15 @@ import TelegramCredentialsInput from './TelegramCredentialsInput'
...
-  hasCredential: initialHasCredential,
-  onNext,
+  setActiveStep,
 }: {
-  hasCredential: boolean
-  onNext: (changes: any, next?: boolean) => void
+  setActiveStep: Dispatch<SetStateAction<TelegramProgress>>
 }) => {
+  const { campaign, setCampaign } = useContext(CampaignContext)
+  const { hasCredential: initialHasCredential, progress } = campaign
```
Comment (lamkeewei) [inline frontend/src/components/dashboard/create/telegram/TelegramCredentials.tsx:39]:
```suggestion
  const { hasCredential: initialHasCredential } = campaign
```

Minor issue, I think `progress` is unused.

File: frontend/src/components/dashboard/create/telegram/TelegramCredentials.tsx [hunk 2]
```diff
@@ -371,10 +370,10 @@ const TelegramCredentials = ({
                 <ErrorBlock>{errorMessage}</ErrorBlock>
               </StepSection>
 
-              <ButtonGroup>
-                <NextButton disabled={!hasCredential} onClick={onNext} />
-                <TextButton onClick={onPrevious}>Previous</TextButton>
-              </ButtonGroup>
+              <NextButton
```
Comment (lamkeewei) [inline frontend/src/components/dashboard/create/telegram/TelegramCredentials.tsx:373]: Previous button is missing here.

File: frontend/src/components/dashboard/create/Create.tsx [hunk 3]
```diff
@@ -23,59 +18,39 @@ const Create = () => {
...
+      setCampaign(campaign)
+      setLoading(false)
+    }
     loadProject(id)
-  }, [id])
+  }, [id, setCampaign])
 
   async function handleFinishLater() {
-    if (campaign.status === Status.Draft) {
+    if (campaign.status === Status.Draft && finishLaterContent) {
```
Comment (lamkeewei) [inline frontend/src/components/dashboard/create/Create.tsx:39]: Minor point. Behaviour for this is slightly changed. The GA event is now only sent for pages that trigger the save draft modal. It doesn't fire for the other pages (credentials, send).

**Comment by jeantanzj** [comment]
82e994d -- this is a very strange commit? dont you have to fix the merge conflicts when you merge develop in

File: frontend/src/components/dashboard/create/Create.tsx (hunk 3 above)
Comment (jeantanzj) [reply frontend/src/components/dashboard/create/Create.tsx:39]: I'll remove this and merge it in

**Review by jeantanzj — APPROVED** [review]

**Pull request merged**